File: multilayers_PhaseDifference_trans.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  3 19:55:58 2022

@author: chihyulee

Translate "multilayers_PhaseDifference_trans.m" from Carlos group

Calculate A,R,T for both states; 
Calculate phase difference of R(Phi) and A(Psi) of different layers 
"""

import numpy as np 
import pandas as pd 
import warnings
import import_refr_in_transfer_matrix_trans as f
import matplotlib.pyplot as plt 
import timeit
import time
%matplotlib widget
import ipywidgets as widgets

 

############### Parameters ################
theta_inc=0       ##Incident angle in degrees
number_layers=5   ##Total number of layers (without counting air)
min_lambda=300    ##Minimum lambda to run simulation
max_lambda=799    ##Maximum lambda to run simulation
n0=1              ##refraction index of incident medium
pol='TE'          ##Choose either TE or TM polarization (relevant under oblique incidence)
PCM='SbSe'        ##Chose the PCM materials 
d_PCM= float(input('PCM thickness:')) #15    ##Thickness of PCM
d_Ag1= float(input('Ag1 thickness:')) #30     ##Thickness of Ag1
d_ITO= float(input('ITO thickness:')) #150
d_Ag2= float(input('Ag2 thickness:')) #100
d_air=1000        ##Thickness of Air 



############### Import all the refractive idexes from the files #############
start = timeit.default_timer()

# Order from top to bottom 
stack_layers=[PCM,'Ag','ITO','Ag','Air']
stack_thickness=[d_PCM, d_Ag1, d_ITO, d_Ag2, d_air] 
stack_thickness_nm=[i *1e-9 for i in stack_thickness]
n_crystalline=np.zeros((max_lambda-min_lambda+1, number_layers))  
k_crystalline=np.zeros((max_lambda-min_lambda+1, number_layers))  
n_amorphous=np.zeros((max_lambda-min_lambda+1, number_layers))    
k_amorphous=np.zeros((max_lambda-min_lambda+1, number_layers))    

for i,l in enumerate(stack_layers):
    
    #materials layer 
    if i is not len(stack_layers)-1:
        n_crystalline[:,i]= f.read_refractive_index_file(l,min_lambda)[1][:max_lambda-min_lambda+1,0]
        k_crystalline[:,i]= f.read_refractive_index_file(l,min_lambda)[1][:max_lambda-min_lambda+1,1]
        n_amorphous[:,i]= f.read_refractive_index_file(l,min_lambda)[1][:max_lambda-min_lambda+1,2]
        k_amorphous[:,i]= f.read_refractive_index_file(l,min_lambda)[1][:max_lambda-min_lambda+1,3]
   #Air layer
    else:
        n_crystalline[:,i]= np.ones(max_lambda-min_lambda+1)
        k_crystalline[:,i]= np.zeros(max_lambda-min_lambda+1)
        n_amorphous[:,i]= np.ones(max_lambda-min_lambda+1)
        k_amorphous[:,i]= np.zeros(max_lambda-min_lambda+1)



############## Transfer Matrix Calculations ##################

#incident angle in radians
theta0= theta_inc*np.pi/180 


#array containing X, A, R, T, Phi, Psi
results_c= f.transfer_matrix_trans(theta0, n0, max_lambda, min_lambda, number_layers, n_crystalline, k_crystalline, stack_thickness_nm, pol)
results_a= f.transfer_matrix_trans(theta0, n0, max_lambda, min_lambda, number_layers, n_amorphous, k_amorphous, stack_thickness_nm, pol)

############## Plot Response after Calculation ##############################

x= results_a[0, :]/1e-9
plt.figure(figsize = (10,6))
plt.xlabel('Wavelength(nm)')
plt.plot(x, results_c[2,:],'-', c='blue', label='$R_{cry}$')
plt.plot(x, results_c[3,:],'--',c= 'red', label='$T_{cry}$')
plt.plot(x, results_a[2,:],'-', c='yellow', label='$R_{amo}$')
plt.plot(x, results_a[3,:],'--',c= 'purple', label='$T_{amo}$')
plt.plot(x, (results_c[-2,:]-results_a[-2,:])/(np.pi*2),'-', c='green', label='$\phi_{R}$' )
plt.plot(x, (results_c[-1,:]-results_a[-1,:])/(np.pi*2),'--', c='c', label='$\psi_{T}$' )
plt.legend(loc=1)

stop = timeit.default_timer()
time= stop-start
materials= zip(stack_layers, stack_thickness)
max_r= np.real(np.max(results_c[2,:]))
max_r_index= np.where(results_c[2,:]==max_r)
plt.title(list(materials))
plt.suptitle(pol +' spectrum Inc: '+ str(theta_inc)+' ${^\circ}$'+
             ','+'Max R: {:.3f}@ {}nm '.format(max_r,np.real(x[max_r_index])[0])+
             ','+ 'Time:{:.3f}s'.format(time))

File: multilayers_PhaseDifference_trans.m
%% THIS PROGRAM CALCULATES A, R and T FOR BOTH SbSe PHASES, and PHASE DIFFERNCE of R(Phi) and T(Psi) OF A STACK OF LAYERS
%% HOW TO USE:
% 1. Input the parameters in the block 1.
% 2. Be sure that the refractive index of the material you want to use is
%    included in the file import_refr_in. 
% 3. Enter the refractive index of each layer in order (in block #3). First entry
%    corresponds to first layer.
% 4. Input the thickness of each layer in block 4. The number of entries must be
%    consistent with the number of layers in parameters.
% 5. Run.
%% By Carlos Rios, August 2015
%% Revised By Yi-Siou Huang, April 2022
clear all
close all
clc

%% 1. PARAMETERS
theta_inc=0;        %% Incident angle in degrees
number_layers=5;    %% Total number of layers (without counting air)
                    %ATTENTION: number_layers must be consistent with the # of entries in d and n,k.
global min_lambda   % Just to be sure min_lambda is taken inside import_refr_in
min_lambda=300;     %% Minimum lambda to run simulation
max_lambda=799;     %% Maximum lambda to run simulation / Beware: some materials n and k are given up to 800nm
n0=1;               %% refraction index of incident medium
pol='TE';           %% Choose either TE or TM polarization (relevant under oblique incidence)

%% 2. IMPORTS ALL THE REFRACTIVE INDEX FROM THE FILE import_refr_in
import_refr_in      
k_SiO2=zeros(1,1700);
n_SiO2=zeros(1,1700);
n_SiO2(1,:)=1.44;

%% 3. REFRACTIVE INDEX OF EACH LAYER first corresponds to upper layer
% This step defines which material goes in which layer
n=struct('n_val',{n_SbSecry, n_Ag, n_ITO, n_Ag, ones(500)});
k=struct('k_val',{k_SbSecry, k_Ag, k_ITO, k_Ag, zeros(500)});

n2=struct('n_val',{n_SbSeam, n_Ag, n_ITO, n_Ag, ones(500)});
k2=struct('k_val',{k_SbSeam, k_Ag, k_ITO, k_Ag, zeros(500)});

%% 4. LAYER THICKNESSES first corresponds to upper layer
d=1e-9*[15,30,150,100,1000];
% d=1e-9*[20,30,120,25,1000]; good for T
% d=1e-9*[10,12,80,100,1000]; good for R

   
%% 5. TRANSFER MATRIX CALCULATIONS
% Phi is phase difference in reflection
% Psi is phase differene in transmission
theta0=theta_inc*pi/180;  % incident angle in radians
[X,A,R,T,Phi,Psi]= transfer_matrix_trans(theta0, n0, max_lambda, min_lambda, number_layers, n, k, d, pol);
[Xam,Aam,Ram,Tam,Phiam,Psiam]= transfer_matrix_trans(theta0, n0, max_lambda, min_lambda, number_layers, n2, k2, d, pol);

%% 6. PLOT
    % ALL DOWN HERE IS TO GENERATE THE PLOT, IT DOES NOTHING TO DO WITH THE
    % CALCULATION ITSELF
         figure, 
         p=plot(X,R,'-',X,T,'--', X,Ram,'-',X,Tam,'--',X,(Phi-Phiam)/(2*pi),'--',X, (Psi-Psiam)/(2*pi),'-')
         set(p,'LineWidth',2);
         xlabel('Wavelength (nm)') 
         legend('Rcry','Tcry','Ram','Tam','Phase diff of R','Phase diff of T') 
%          legend('R','T','A', 'Ram','Tam','Aam') 
         axis([min(X) max(X) -1 1])
         
         
        imax = find(max(R) == R);
        text(min_lambda,0,['Maximum R=  ',num2str(R(imax)),' @ ',num2str(X(imax)),'nm',],...
        'VerticalAlignment','bottom',...
        'HorizontalAlignment','left',...
        'FontSize',8)
        text(min_lambda,0,['Layers=',mat2str(d*1e9),' nm   ','Angle=', mat2str(theta0*180/pi)],...
        'VerticalAlignment','top',...
        'HorizontalAlignment','left',...
        'FontSize',8);
        title(strcat(pol,' spectrum',' Inc: ', num2str(theta_inc),'^{o}'));

        
        %num=num2str(150*10^9);
        %saveas(gcf,['ITO_',num,'nm'],'bmp');
